src/common/tensor_inspector.h

```diff
@@ -0,0 +1,800 @@
+      }
+      tensor_info_to_string<DType>(&std::cout);
+      std::cout << "To print a part of the tensor," <<
+	        " please specify a position, seperated by \",\"" << std::endl;
+      std::cout << "\"e\" for the entire tensor, " <<
+          "\"d\" to dump value to file, " <<
+          "\"b\" to break, " <<
+          "\"s\" to skip all: ";
+      std::string str;
+      std::cin >> str;
+      if (str == "b") {
+        break;
+      } else if (str == "e") {
+        to_string_helper<DType>(&std::cout);
+        continue;
+      } else if (str == "s") {
+        InspectorManager::get()->interactive_print_skip_all_ = true;
+        break;
+      } else if (str == "d") {
+        while (true) {
+          std::cout << "Please enter a tag: ";
+          std::cin >> str;
+          if (str.find(' ') != std::string::npos) {
+            std::cout << "Invalid input. ";
+            continue;
+          }
+          dump_to_file_helper<DType>(str);
+          break;
+        }
+        continue;
+      }
+      std::vector<index_t> pos;
+      if (parse_position(&pos, str)) {
+        std::vector<index_t> sub_shape;
+        index_t offset;
+        print_locator(pos, &sub_shape, &offset);
+        to_string_helper<DType>(&std::cout, sub_shape, offset);
+      } else {
+        std::cout << "invalid input" << std::endl;
+      }
+    }
+  }
+
+  /*!
+   * \brief build the lambda function, aka the checker, given its type
+   * \tparam DType the data type
+   * \param ct the type of the checker
+   */
+  template<typename DType>
+  std::function<bool(DType)> get_checker(CheckerType ct) {
+    switch (ct) {
+      case NegativeChecker:
+        return [] (DType x) {
+              return x < 0;
+            };
+      case PositiveChecker:
+        return [] (DType x) {
+              return x > 0;
+            };
+      case ZeroChecker:
+        return [] (DType x) {
+              return x == 0;
+            };
+      case NaNChecker:
+        if (std::is_same<DType, float>::value || std::is_same<DType, double>::value ||
+            std::is_same<DType, mshadow::half::half_t>::value) {
+          return [] (DType x) {
+                return x != x;
+              };
+        } else {
+          LOG(WARNING) << "NaNChecker only applies to float types. " <<
+              "Lambda will always return false.";
+        }
+        break;
+      case InfChecker:
+        if (std::is_same<DType, float>::value || std::is_same<DType, double>::value ||
+            std::is_same<DType, mshadow::half::half_t>::value) {
+          return [] (DType x) {
+                return x == (DType)1.0 / 0.0f || x == -(DType)1.0 / 0.0f;
+              };
+        } else {
+          LOG(WARNING) << "InfChecker only applies to float types. " <<
+              "Lambda will always return false.";
+        }
+        break;
+      case PositiveInfChecker:
+        if (std::is_same<DType, float>::value || std::is_same<DType, double>::value ||
```

Comment:
Can we just use std::isinf?

Reply:
The issue is, get_checker is templated, and there is no isinf() for integer types. The code won't compile then

Reply:
sadly we are not using c++ 17. otherwise we could do static branching and use isinf()